src/python/pants/backend/docker/value_interpolation.py

```diff
@@ -62,14 +57,68 @@ class DockerBuildArgsInterpolationValue(DockerInterpolationValue):
     _attribute_error_type = DockerBuildArgInterpolationError
 
 
-class DockerInterpolationContext(FrozenDict[str, DockerInterpolationValue]):
+class DockerInterpolationContext(FrozenDict[str, Union[str, DockerInterpolationValue]]):
     @classmethod
-    def from_dict(
-        cls, data: Mapping[str, Mapping[str, str] | DockerInterpolationValue]
-    ) -> DockerInterpolationContext:
+    def from_dict(cls, data: Mapping[str, str | Mapping[str, str]]) -> DockerInterpolationContext:
         return DockerInterpolationContext(
-            {key: DockerInterpolationValue.create(value) for key, value in data.items()}
+            {key: cls.create_value(value) for key, value in data.items()}
         )
 
-    def merge(self, other: Mapping[str, Mapping[str, str]]) -> DockerInterpolationContext:
+    @staticmethod
+    def create_value(value: str | Mapping[str, str]) -> str | DockerInterpolationValue:
+        """Ensure that `value` satisfies the type `DockerInterpolationValue`."""
+        if isinstance(value, (str, DockerInterpolationValue)):
+            return value
+        return DockerInterpolationValue(value)
+
+    def merge(self, other: Mapping[str, str | Mapping[str, str]]) -> DockerInterpolationContext:
         return DockerInterpolationContext.from_dict({**self, **other})
+
+    def format(
+        self, text: str, *, source: TextSource, error_cls: type[ErrorT] | None = None
+    ) -> str:
+        try:
+            return text.format(**self)
+        except (KeyError, DockerInterpolationError) as e:
+            default_error_cls = DockerInterpolationError
+            msg = f"Invalid value for the {source}: {text!r}.\n\n"
+            if isinstance(e, DockerInterpolationError):
+                default_error_cls = type(e)
+                msg += str(e)
+            else:
+                # KeyError
+                msg += f"The placeholder {e} is unknown."
+                if self:
+                    msg += f" Try with one of: {', '.join(sorted(self.keys()))}."
+                else:
+                    msg += (
+                        " There are currently no known placeholders to use. These placeholders "
+                        "can come from `[docker].build_args` or parsed from instructions in your "
+                        "`Dockerfile`."
+                    )
+            raise (error_cls or default_error_cls)(msg) from e
+
+    @dataclass(frozen=True)
+    class TextSource:
+        address: Address | None = None
+        target_alias: str | None = None
+        field_alias: str | None = None
+        options_scope: str | None = None
```

Comment:
Not sure how I feel about having an inner class like this. Is your motivation namespacing? It can be worth it, but does make things more complex.

Reply:
more complex how?

Namespacing is part of it. What I did consider briefly was to have a separate class doing the formatting (kind of a mix of `TestSource` and `DockerInterpolationContext.format`.

The other part is naming and importing. With this, I only have to import the `DockerInterpolationContext`. And the `TextSource` is just an alternative to having a bunch of kwargs on the `format` method.. as I have two different ways to present where the value being formatted comes from, and didn't want to leave any formatting outside of this module here..